File: client/src/pairingHeap.js
class PairingHeapNode {
  constructor(key, data) {
    this.key = key;
    this.data = data;
    this.parent = null;
    this.children = [];
  }
}

export default class PairingHeap {
  constructor() {
    this.root = null;
    this.lst = [];
  }

  is_empty() {
    return this.root === null;
  }

  insert(key, data) {
    const node = new PairingHeapNode(key,data);          // Create a new node with the given key
    this.root = this.merge(this.root, node);        // Merge the new node with the root node of the heap
  }

  find_min() {
    if (this.is_empty()) {
      throw new Error('Heap is empty');             // Checks for empty heap
    }
    return this.root.data;                           // Returns the key of the root node
  }

  delete_min() {
    if (this.is_empty()) {
      throw new Error('Heap is empty');             // Checks for empty heap
    }
    const min_node = this.root;                     // Saving as a temp.
    if (min_node.children.length > 0) {
      this.root = this.pairwise_merge(min_node.children);  // Merge the children of the minimum node pairwise
    } else {
      this.root = null;                             // If the minimum node has no children, set the root to null
    }

    return min_node.data ;                     // Returns key of minimum node
  }

  merge(heap1, heap2) {
    // Check if either heap is null
    if (heap1 === null) {
      return heap2;
    }
    if (heap2 === null) {
      return heap1;
    }
    // Make heap1 the root of the merged heap
    if (heap1.key < heap2.key) {
      heap2.parent = heap1;
      heap1.children.push(heap2);
      return heap1;
    } else {
      // Make heap2 the root of the merged heap
      heap1.parent = heap2;
      heap2.children.push(heap1);
      return heap2;
    }
  }

  allElements() {
    const elements = [];
    if (this.is_empty()) {
      return elements;   // Return empty array for an empty heap
    }
    const stack = [this.root];
    while (stack.length > 0) {
      const node = stack.pop();
      elements.push(node.data);
      for (let i = node.children.length - 1; i >= 0; i--) {
        stack.push(node.children[i]);
      }
    }
    return elements;
  }

  pairwise_merge(nodes) {
    // Base case: if there is only one node, return it
    if (nodes.length === 1) {
      return nodes[0];
    } else {
      // Merge nodes pairwise using the merge method
      const new_nodes = [];
      for (let i = 0; i < nodes.length - 1; i += 2) {
        const new_node = this.merge(nodes[i], nodes[i+1]);
        new_nodes.push(new_node);
      }
      // If there is an odd number of nodes, append the last one to the new list
      if (nodes.length % 2 === 1) {
        new_nodes.push(nodes[nodes.length-1]);
      }
      // Recursively merge the new list of nodes
      return this.pairwise_merge(new_nodes);
    }
  }
  
  getAll() {
    const elements = [];
    if (this.is_empty()) {
      return elements;   // Return empty array for an empty heap
    }
    const stack = [this.root];
    while (stack.length > 0) {
      const node = stack.pop();
      elements.push(node);
      for (let i = node.children.length - 1; i >= 0; i--) {
        stack.push(node.children[i]);
      }
    }
    elements.sort((a, b) => a.key - b.key); // sort the elements in ascending order
    return elements.map(node => node.data); // return an array of sorted node data
}

}


  
  
  // let heap = new PairingHeap();
  
  // heap.insert(5);
  // heap.insert(30);
  // heap.insert(17);
  // heap.insert(1);
  // heap.insert(1291);
  // heap.insert(0.1);
  // heap.insert(1.1);
  
  
  
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // heap.insert(55);
  // heap.insert(3);
  // heap.insert(330)
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  // console.log(heap.find_min(),heap.delete_min()); 
  

File: Performance Comparison/pairing_heap.py
# class PairingHeapNode:
#     def __init__(self, key):
#         self.key = key
#         self.parent = None
#         self.children = []

# class PairingHeap:
#     def __init__(self):
#         self.root = None

#     def is_empty(self):
#         return self.root is None

#     def insert(self, key):
#         node = PairingHeapNode(key)                 # Create a new node with the given key
#         self.root = self.merge(self.root, node)     # Merge the new node with the root node of the heap

#     def find_min(self):
#         if self.is_empty():
#             raise Exception('Heap is empty')        #checks for empty heap

#         return self.root.key                        #Returns the key of the root node.

#     def delete_min(self):
#         if self.is_empty():
#             raise Exception('Heap is empty')        #checks for empty heap
#         min_node = self.root                        #saving as a temp.

#         if len(min_node.children) > 0:
#             self.root = self.pairwise_merge(min_node.children) # Merge the children of the minimum node pairwise
#         else:
#             self.root = None            #If the minimum node has no children, set the root to None

#         return min_node.key             #Returns key of minimum node.

#     def merge(self, heap1, heap2):
#         # Check if either heap is None
#         if heap1 is None:
#             return heap2
#         if heap2 is None:
#             return heap1
#         # Make heap1 the root of the merged heap
#         if heap1.key < heap2.key:
#             heap2.parent = heap1
#             heap1.children.append(heap2)
#             return heap1
#         else:
#             # Make heap2 the root of the merged heap
#             heap1.parent = heap2
#             heap2.children.append(heap1)
#             return heap2


#     def pairwise_merge(self, nodes):
#         # Base case: if there is only one node, return it
#         if len(nodes) == 1:
#             return nodes[0]
#         else:
#             # Merge nodes pairwise using the merge method
#             new_nodes = []
#             for i in range(0, len(nodes) - 1, 2):
#                 new_node = self.merge(nodes[i], nodes[i+1])
#                 new_nodes.append(new_node)
#             # If there is an odd number of nodes, append the last one to the new list
#             if len(nodes) % 2 == 1:
#                 new_nodes.append(nodes[-1])
#             # Recursively merge the new list of nodes
#             return self.pairwise_merge(new_nodes)


class PairingHeap:
    class Node:
        def __init__(self, value):
            self.value = value
            self.children = []
            self.parent = None

    def __init__(self):
        self.root = None
        self.size = 0

    def insert(self, value):
        new_node = PairingHeap.Node(value)
        self.root = self.merge(self.root, new_node)
        self.size += 1

    def pop(self):
        if not self.root:
            return None

        popped_value = self.root.value

        # Merge all children of the root node
        if self.root.children:
            self.root = self.merge_pairs(self.root.children)
        else:
            self.root = None
        self.size -= 1

        return popped_value

    def merge(self, node1, node2):
        if not node1:
            return node2
        elif not node2:
            return node1
        elif node1.value < node2.value:
            node1.children.append(node2)
            node2.parent = node1
            return node1
        else:
            node2.children.append(node1)
            node1.parent = node2
            return node2

    def merge_pairs(self, nodes):
        if len(nodes) == 1:
            return nodes[0]

        # Pairwise merge adjacent nodes and then merge the resulting pairs
        pairs = [
            self.merge(nodes[i], nodes[i + 1]) for i in range(0, len(nodes) - 1, 2)
        ]
        if len(nodes) % 2 == 1:
            pairs.append(nodes[-1])
        return self.merge_pairs(pairs)
